libevmasm/Assembly.cpp

```diff
@@ -1156,6 +1272,166 @@ LinkerObject const& Assembly::assemble() const
+			{
+				case Operation:
+					ret.bytecode.push_back(static_cast<uint8_t>(i.instruction()));
+					break;
+				case Push:
+				{
+					unsigned b = numberEncodingSize(i.data());
+					if (b == 0 && !m_evmVersion.hasPush0())
+					{
+						b = 1;
+					}
+					ret.bytecode.push_back(static_cast<uint8_t>(pushInstruction(b)));
+					if (b > 0)
+					{
+						ret.bytecode.resize(ret.bytecode.size() + b);
+						bytesRef byr(&ret.bytecode.back() + 1 - b, b);
+						toBigEndian(i.data(), byr);
+					}
+					break;
+				}
+				case PushLibraryAddress:
+					ret.bytecode.push_back(static_cast<uint8_t>(Instruction::PUSH20));
+					ret.linkReferences[ret.bytecode.size()] = m_libraries.at(i.data());
+					ret.bytecode.resize(ret.bytecode.size() + 20);
+					break;
+				case VerbatimBytecode:
+					ret.bytecode += i.verbatimData();
+					break;
+				case PushDeployTimeAddress:
+					ret.bytecode.push_back(static_cast<uint8_t>(Instruction::PUSH20));
+					ret.bytecode.resize(ret.bytecode.size() + 20);
+					break;
+				case Tag:
+				{
+					assertThrow(i.data() != 0, AssemblyException, "Invalid tag position.");
+					assertThrow(i.splitForeignPushTag().first == std::numeric_limits<size_t>::max(), AssemblyException, "Foreign tag.");
+					size_t tagId = static_cast<size_t>(i.data());
+					assertThrow(ret.bytecode.size() < 0xffffffffL, AssemblyException, "Tag too large.");
+					assertThrow(m_tagPositionsInBytecode[tagId] == std::numeric_limits<size_t>::max(), AssemblyException, "Duplicate tag position.");
+					m_tagPositionsInBytecode[tagId] = ret.bytecode.size();
+					break;
+				}
+				default:
+					assertThrow(false, InvalidOpcode, "Unexpected opcode while assembling.");
+			}
+		}
+
+		auto sectionEnd = ret.bytecode.size();
+		setCodeSectionSize(codeSectionIndex, sectionEnd - sectionStart);
+	}
+
+	for (auto const& [name, tagInfo]: m_namedTags)
+	{
+		size_t position = m_tagPositionsInBytecode.at(tagInfo.id);
+		std::optional<size_t> tagIndex;
+		for (auto& codeSection: m_codeSections)
+			for (auto&& [index, item]: codeSection.items | ranges::views::enumerate)
+				if (item.type() == Tag && static_cast<size_t>(item.data()) == tagInfo.id)
+				{
+					tagIndex = index;
+					break;
+				}
+		ret.functionDebugData[name] = {
+			position == std::numeric_limits<size_t>::max() ? std::nullopt : std::optional<size_t>{position},
+			tagIndex,
+			tagInfo.sourceID,
+			tagInfo.params,
+			tagInfo.returns
+		};
+	}
+
+	// Fill the num_container_sections
+	auto const numContainers = containerSizes.size();
+	if (numContainers > 0)
+	{
+		bytes containerSectionHeader = {};
+		containerSectionHeader.push_back(0x03);
+		appendBigEndianUint16(containerSectionHeader, numContainers);
+
+		for (auto s: containerSizes)
+			appendBigEndianUint16(containerSectionHeader, s.second);
+
+		ret.bytecode.insert(ret.bytecode.begin() + static_cast<ptrdiff_t>(startOfContainerSectionHeader),
+							containerSectionHeader.begin(), containerSectionHeader.end());
```

Comment:
Yes and No. :) It depends. In this context it looks like we know all containers before but we don't really know. We have to add there only referenced containers but they can be a proper subset of `m_subs`. This is deferred because in later PR when implementing instructions which reference sub containers I add a map which store all them and fill the section accordingly. On the other hand I think unreferenced sub containers should be removed on some earlier compilation stage. Similarly as it's done for functions. 
Formatting fixed.

Reply:
If extra pass is acceptable I will encapsulate this into a function and do it at the beginning. But to make it possible I need to introduce instructions which refers  sub containers. So for now I will just make this function returning `m_subs` or  empty array.

Reply:
Yeah, of course. I'm not expecting new functionality here, just wanted the existing structure to be rearranged.